native-addons-are-close-to-being-on-par-with-js-modules.md

```diff
@@ -0,0 +1,113 @@
+
+[N-API][] was introduced to address the problem of the binary compatibility of native addons with versions of Node.js against which they were build *as well as all subsequent versions*. It abstracts away essential JavaScript engine and Node.js APIs, presenting them to native addon maintainers as a collection of ABI-stable APIs. The stability of N-API is assured by the fact that all APIs are C APIs rather than C++ APIs and by a commitment on the part of the Node.js core developer community to avoid modifying the interface in an incompatible way.
+
+N-API eliminates one of the three variables that causes the proliferation of binaries required to provide a single Node.js native addon: the integer value that represents the Node.js version against which the binary is designed to work (`NODE_MODULE_VERSION`).
+
+### 4. Multiple Loading
+The advent of worker threads in Node.js, and even earlier modules such as [`vm`][], make it possible to load modules multiple times. This is easy to do with JavaScript modules because Node.js simply re-reads and re-interprets the file defining the JavaScript module and the resulting object resides in a separate context, without inadvertently sharing any state with its previously loaded incarnation. In contrast, it has only recently become possible to load native addons more than once. The reason for this is architectural. In particular, it is because the mechanism for loading a native addon was as follows:
+
+0. Node.js calculates the absolute path to the native addon.
+0. Node.js calls the method [`process.dlopen()`][], passing in the absolute path previously calculated.
+0. On the native side, Node.js calls [`dlopen(3)`][] or [`uv_dlopen()`][] on Windows.
+0. The addon is loaded and it executes a so-called DSO constructor function as part of the loading process. The function passes a pointer to a structure of type `node::nm_module` to Node.js.
+0. The `node::node_module::nm_register_func` or `node::node_module::nm_context_register_func` member of the structure holds a pointer to a function which is responsible for populating the object which will become the result of having loaded the module (`module.exports`).
+
+The problem with DSO constructor functions is that they only run the first time an addon is loaded into memory. On subsequent occasions, `dlopen(3)` or `uv_dlopen()` will short-circuit, returning a reference to an already loaded addon, and the function will not run. Thus, subsequent attempts to load the addon fail.
+
+The solution to this problem was introduced in [3828fc62][]. In addition to relying on a DSO constructor, as a fallback Node.js now looks for a well-known symbol exported by the module: `node_register_module_v<number>`, where `<number>` is an integer representing the version of the addon. It only does so if the DSO constructor fails to run. The well-known symbols is the address of the module initialization function that is also stored in `node::node_module::nm_context_register_func`.
```

Comment:
The link `[3828fc62][]` appears broken.

Reply:
@jschlight FWIW if you open the actual blog post on GitHub rather than this diff-view of it, the link does work: https://github.com/gabrielschulhof/gabrielschulhof.github.com/blob/blog-post/native-addons-are-close-to-being-on-par-with-js-modules.md#4-multiple-loading